src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -21,47 +24,126 @@ internal class FileShareProvider: IArtifactProvider
         {
+            BlobStoreClientTelemetry clientTelemetry;
+            DedupManifestArtifactClient dedupManifestClient = DedupManifestArtifactClientFactory.CreateDedupManifestClient(context, connection, cancellationToken, out clientTelemetry);
+            using (clientTelemetry)
+            {
+                FileShareActionRecord downloadRecord = clientTelemetry.CreateRecord<FileShareActionRecord>((level, uri, type) =>
+                    new FileShareActionRecord(level, uri, type, nameof(CopyFileShareAsync), context));
+
+                await clientTelemetry.MeasureActionAsync(
+                    record: downloadRecord,
+                    actionAsync: async () =>
+                    {
+                        return await DownloadArtifactsAsync(downloadParameters, buildArtifacts, cancellationToken);
+                    }
+                );
+
+                // Send results to CustomerIntelligence
+                context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineArtifact, record: downloadRecord);
+            }
+        }
+
+        public async Task<FileShareDownloadResult> DownloadArtifactsAsync(PipelineArtifactDownloadParameters downloadParameters, IEnumerable<BuildArtifact> buildArtifacts, CancellationToken cancellationToken)
+        {
+            var fileShareDownloadResult = new FileShareDownloadResult(0, 0, 0);
+
             foreach (var buildArtifact in buildArtifacts)
             {
                 var downloadRootPath = Path.Combine(buildArtifact.Resource.Data, buildArtifact.Name);
                 var minimatchPatterns = downloadParameters.MinimatchFilters.Select(pattern => Path.Combine(buildArtifact.Resource.Data, pattern));
-                await this.CopyFileShareAsync(downloadRootPath, Path.Combine(downloadParameters.TargetDirectory, buildArtifact.Name), minimatchPatterns, cancellationToken);
+                await this.CopyFileShareAsync(downloadRootPath, Path.Combine(downloadParameters.TargetDirectory, buildArtifact.Name), minimatchPatterns, cancellationToken, fileShareDownloadResult);
+            }
+            
+            return fileShareDownloadResult;
+        }
+
+        public async Task PublishSingleArtifactAsync(
+            string sourcePath,
+            string destPath,
+            int parallelCount,
+            CancellationToken cancellationToken) 
+        {
+            BlobStoreClientTelemetry clientTelemetry;
+            DedupManifestArtifactClient dedupManifestClient = DedupManifestArtifactClientFactory.CreateDedupManifestClient(context, connection, cancellationToken, out clientTelemetry);
+            using (clientTelemetry)
+            {
+                FileShareActionRecord downloadRecord = clientTelemetry.CreateRecord<FileShareActionRecord>((level, uri, type) =>
```

Comment:
>downloadRecord [](start = 38, length = 14)

"downloadRecord" is not the right name for the context, should be "pubishRecord"